### PR TITLE
Key & Account Queues

### PR DESCRIPTION
**Pull request by tonyhb**
See the spec.  More details to follow.

_(nb. PR body & title rewritten by @cdzombak 2024-09-03; these are not @tonyhb's words)_

## Issues

- See: https://linear.app/inngest/project/key-and-account-level-queues-26f88c23df90/issues

## Type of change (choose one)
- [x] Chore (refactors, upgrades, etc.)
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Security fix (non-breaking change that fixes a potential vulnerability)
- [ ] Docs
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)

### CODE REVIEW

File: pkg/execution/state/redis_state/queue.go [hunk 1]
```diff
@@ -564,42 +565,82 @@ func (q QueueShard) Partition() string {
...
+	// function ID or the environment scope itself.
+	EnvID *uuid.UUID `json:"wsID,omitempty"`
+	// AccountID represents the account ID for the partition.  This ONLY exists
+	// if the partition represents an account-level concurrency key.
+	AccountID *uuid.UUID `json:"aID,omitempty"`
+	// Priority represents the partition's priority.  This currently exists
+	// on the partition (instead of fn metadata) to reduce lookups when peeking
+	// many partitions.
 	Priority uint `json:"p"`
-	Paused   bool `json:"off"`
```
Comment (cdzombak) [inline pkg/execution/state/redis_state/queue.go:576]: nb. requires app migration for deploy

File: pkg/execution/state/state.go [hunk 2]
```diff
@@ -104,6 +105,50 @@ type CustomConcurrency struct {
 	Limit int `json:"l"`
 }
 
+func (c CustomConcurrency) Validate() error {
+	// Keys must always be in the format of "$prefix:$id:$key", in which prefix
+	// is one of "f" | "e" | "a", depending on function, env, or account-level scopes.
```
Comment (cdzombak) [inline pkg/execution/state/state.go:110]: can make this function more thorough

File: pkg/enums/partition_type.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
+//go:generate go run github.com/dmarkham/enumer -trimprefix=PartitionType -type=PartitionType -json -text -gqlgen
```
Comment (cdzombak) [inline pkg/enums/partition_type.go:1]:
```suggestion
//go:generate go run github.com/dmarkham/enumer -trimprefix=PartitionType -type=PartitionType -json -text
```

File: pkg/execution/state/state.go [hunk 4]
```diff
@@ -104,6 +105,50 @@ type CustomConcurrency struct {
...
+	if len(parts) != 3 {
+		// Invalid by default
+		return enums.ConcurrencyScopeFn, id, err
+	}
+
+	id, err = uuid.Parse(parts[1])
+
+	switch parts[0] {
+	case "f":
+		return enums.ConcurrencyScopeFn, id, err
```
Comment (cdzombak) [inline pkg/execution/state/state.go:142]: would like to see concurrency scope ints be explicitly defined, and also we can remove `gqlgen`

File: pkg/execution/state/state.go [hunk 5]
```diff
@@ -104,6 +105,50 @@ type CustomConcurrency struct {
...
+	}
+	switch c.Key[0] {
+	case 'f', 'e', 'a':
+	default:
+		return fmt.Errorf("unknown concurrency key scope: %v", c.Key[0])
+	}
+	return nil
+}
+
+func (c CustomConcurrency) ParseKey() (scope enums.ConcurrencyScope, id uuid.UUID, err error) {
```
Comment (cdzombak) [inline pkg/execution/state/state.go:122]: want: test coverage

File: pkg/execution/state/state.go [hunk 6]
```diff
@@ -104,6 +105,50 @@ type CustomConcurrency struct {
 	Limit int `json:"l"`
 }
 
+func (c CustomConcurrency) Validate() error {
```
Comment (cdzombak) [inline pkg/execution/state/state.go:108]: want: test coverage

File: pkg/execution/state/redis_state/queue.go (hunk 1 above)
Comment (cdzombak) [reply pkg/execution/state/redis_state/queue.go:576]: https://linear.app/inngest/issue/INN-3236/migrations-for-partition-paused-→-fn-paused

File: pkg/execution/state/state.go (hunk 4 above)
Comment (cdzombak) [reply pkg/execution/state/state.go:142]: ✅ 9ff75544f71439c92e04d11d11965a4132a49c03

File: pkg/execution/state/state.go (hunk 5 above)
Comment (cdzombak) [reply pkg/execution/state/state.go:122]: ✅ 41619f07fad20338e21dd2e7f5acc786f084047d

File: pkg/execution/state/state.go (hunk 6 above)
Comment (cdzombak) [reply pkg/execution/state/state.go:108]: (We'll do this later.)

File: pkg/execution/state/redis_state/queue.go (hunk 1 above)
Comment (tonyhb) [reply pkg/execution/state/redis_state/queue.go:576]: Done.  Thanks!

File: pkg/execution/state/state.go (hunk 6 above)
Comment (cdzombak) [reply pkg/execution/state/state.go:110]: Moved to https://linear.app/inngest/issue/INN-3249/improve-test-coverage-for-customconcurrencyvalidate

**Review by cdzombak — APPROVED** [review]

**Pull request merged**